### DISCUSSION TITLE
Any plan to support a folding scheme?

### DISCUSSION THREAD

**Discussion by weijiguo** [Consensys/gnark#1133] [General]
For example, the sangria scheme for Plonk. This should be very useful and powerful in scenarios of IVC.

Searching around I found some PRs and commits mentioning folding. However, as a feature it does not seem to have been planned.

**Comment by ivokub** [comment]
We currently do not have direct plan for implementing folding based schemes, but won't exclude anything if it makes sense. It would be really nice if the folding scheme works directly with PLONK or R1CS arithmetization as it would be quite a major issue to add new arithmetization. We haven't really though about generalizing to CCS either, but maybe would be good addition if shows good initial results.

In general, we're looking into sumcheck based approaches (GKR, Testudo, HyperPLONK), STARKs (PLONK+FRI which we used to have but have disabled for now due to bugs; I have heard people interested in implementing PLONK+STIR), GPU acceleration (already included for Groth16 and for PLONK is also in the works) etc.

We have also briefly thought about using custom gates, but have currently decided against it as it would be really difficult to implement on the frontend in a user-friendly way (and it also leads to performance bottlenecks on the backend).

**Comment by weijiguo** [comment]
Great. That's good to know. Thanks.
